binderhub/health.py

```diff
@@ -0,0 +1,92 @@
+import asyncio
+import json
+
+from functools import wraps
+
+from tornado.httpclient import AsyncHTTPClient
+
+from .base import BaseHandler
+
+
+def retry(_f=None, *, delay=1, attempts=3):
+    """Retry calling the decorated function if it raises an exception
+
+    Repeated calls are spaced by `delay` seconds and a total of `attempts`
+    retries will be made.
+    """
+
+    def repeater(f):
+        @wraps(f)
+        async def wrapper(*args, **kwargs):
+            nonlocal attempts
+            while attempts > 0:
+                try:
+                    return await f(*args, **kwargs)
+                except Exception as e:
+                    if attempts == 1:
+                        raise
+                    else:
+                        attempts -= 1
+                        await asyncio.sleep(delay)
+
+        return wrapper
+
+    if _f is None:
+        return repeater
+    else:
+        return repeater(_f)
+
+
+def false_if_raises(f):
+    """Return False if `f` raises an exception"""
+
+    @wraps(f)
+    async def wrapper(*args, **kwargs):
+        try:
+            res = await f(*args, **kwargs)
+        except Exception as e:
+            res = False
+        return res
+
+    return wrapper
+
+
+class HealthHandler(BaseHandler):
+    """Serve health status"""
+
+    def initialize(self, hub_url=None):
+        self.hub_url = hub_url
+
+    @false_if_raises
+    @retry
+    async def check_jupyterhub_api(self, hub_url):
+        """Check JupyterHub API health"""
+        await AsyncHTTPClient().fetch(hub_url + "hub/health", request_timeout=2)
+
+        return True
+
+    @false_if_raises
+    @retry
+    async def check_docker_registry(self):
+        """Check docker registry health"""
```

Comment:
Note to self: maybe docker registries have a `/health` end point we can call instead?

Reply:
`curl "https://dlqpwel7.gra5.container-registry.ovh.net/api/health" -H "accept: application/json"`

Reply:
Maybe https://dlqpwel7.gra5.container-registry.ovh.net/v2/ and https://gcr.io/v2/ are the better choices as they are part of the docker registry. https://github.com/docker/distribution/issues/629 suggests that you should get a status 401 here if everything is well.